crates/deltalake-core/src/protocol/mod.rs

```diff
@@ -640,6 +640,190 @@ pub struct Protocol {
     /// Minimum version of the Delta write protocol a client must implement to correctly read the
     /// table.
     pub min_writer_version: i32,
+    /// Table features are missing from older versions
+    /// The table features this reader supports
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub reader_features: Option<HashSet<ReaderFeatures>>,
+    /// Table features are missing from older versions
+    /// The table features this writer supports
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub writer_features: Option<HashSet<WriterFeatures>>,
+}
+
+/// Features table readers can support as well as let users know
+/// what is supported
+#[derive(Serialize, Deserialize, Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum ReaderFeatures {
+    /// Mapping of one column to another
+    #[serde(alias = "columnMapping")]
+    COLUMN_MAPPING,
+    /// Deletion vectors for merge, update, delete
+    #[serde(alias = "deletionVectors")]
+    DELETION_VECTORS,
+    /// timestamps without timezone support
+    #[serde(alias = "timestampNtz")]
+    TIMESTAMP_WITHOUT_TIMEZONE,
+    /// version 2 of checkpointing
+    #[serde(alias = "v2Checkpoint")]
+    V2_CHECKPOINT,
+    /// If we do not match any other reader features
+    #[serde(other)]
+    OTHER,
```

Comment:
Should we capture this value using an untagged variant instead? 

Something like this

```rust
#[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
pub enum Example {
    Foo,
    Bar,
    #[serde(untagged)]
    Unknown(String),
}

#[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
struct Features {
    reader: Vec<Example>
}

fn main() {
    let features = Features { 
        reader: vec![Example::Foo, Example::Bar, Example::Unknown("Hello".to_string())],
    };
    println!("{}", serde_json::to_string(&features).unwrap());
    let features: Features = serde_json::from_str(r#"{"reader":["Foo","Bar","Hello"]}"#).unwrap();
    println!("{:#?}", features);
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ef8134eeaf36d87ca8fbb38b57d18765

Reply:
Actually read my mind, I like this more than other.